File: gameserver/src/main/java/l2s/gameserver/network/authcomm/as2gs/AuthResponse.java
package l2s.gameserver.network.authcomm.as2gs;

import io.netty.buffer.ByteBuf;
import l2s.gameserver.network.authcomm.AuthServerClientService;
import l2s.gameserver.network.authcomm.ReceivablePacket;
import l2s.gameserver.network.authcomm.gs2as.OnlineStatus;
import l2s.gameserver.network.authcomm.gs2as.PlayerInGame;
import l2s.gameserver.network.authcomm.vertx.AuthServerConnection;

import java.util.ArrayList;
import java.util.List;

public class AuthResponse extends ReceivablePacket {
    private final List<ServerInfo> _servers;

    public AuthResponse(AuthServerConnection client, ByteBuf byteBuf) {
        super(client, byteBuf);

        int serverId = byteBuf.readUnsignedByte();
        String serverName = readString(byteBuf);
        if (!isReadable()) {
            _servers = new ArrayList<>(1);
            _servers.add(new ServerInfo(serverId, serverName));
        } else {
            int serversCount = byteBuf.readUnsignedByte();
            _servers = new ArrayList<>(serversCount);
            for (int i = 0; i < serversCount; ++i)
                _servers.add(new ServerInfo(byteBuf.readUnsignedByte(), readString(byteBuf)));
        }
    }

    @Override
    protected void runImpl(AuthServerConnection client) {
        client.setAuthed(true);

        for (ServerInfo info : _servers)
            LOGGER.info("Registered on authserver as {} [{}]", info.getId(), info.getName());
        client.sendPacket(new OnlineStatus(true));

        String[] accounts = AuthServerClientService.INSTANCE.getAccounts();
        for (String account : accounts)
            client.sendPacket(new PlayerInGame(account));
    }

    private static class ServerInfo {
        private final int _id;
        private final String _name;

        public ServerInfo(int id, String name) {
            _id = id;
            _name = name;
        }

        public int getId() {
            return _id;
        }

        public String getName() {
            return _name;
        }
    }
}

File: gameserver/src/main/java/l2s/gameserver/network/authcomm/as2gs/AuthResponse.kt
package l2s.gameserver.network.authcomm.as2gs

import io.netty.buffer.ByteBuf
import l2s.gameserver.network.authcomm.AuthServerClientService
import l2s.gameserver.network.authcomm.ReceivablePacket
import l2s.gameserver.network.authcomm.gs2as.OnlineStatus
import l2s.gameserver.network.authcomm.gs2as.PlayerInGame
import l2s.gameserver.network.authcomm.vertx.AuthServerConnection
import org.slf4j.LoggerFactory
import java.util.*

/**
 * @reworked by Bonux
 */
class AuthResponse(client: AuthServerConnection, byteBuf: ByteBuf) : ReceivablePacket(client, byteBuf) {

    private class ServerInfo(val id: Int, val name: String)

    private val _servers: List<ServerInfo>

    init {
        val serverId: Int = byteBuf.readUnsignedByte().toInt()
        val serverName: String = readString(byteBuf)
        if (!isReadable) {
            _servers = listOf(ServerInfo(serverId, serverName))
        } else {
            val serversCount: Int = byteBuf.readUnsignedByte().toInt()
            _servers = ArrayList(serversCount)
            for (i in 0 until serversCount)
                _servers.add(ServerInfo(byteBuf.readUnsignedByte().toInt(), readString(byteBuf)))
        }
    }

    override fun runImpl(client: AuthServerConnection) {
        for (info in _servers) _log.info("Registered on authserver as " + info.id.toString() + " [" + info.name + "]")
        client.sendPacket(OnlineStatus(true))
        val accounts: Array<String> = AuthServerClientService.getAccounts()
        for (account in accounts)
            client.sendPacket(PlayerInGame(account))
    }

    companion object {
        private val _log = LoggerFactory.getLogger(AuthResponse::class.java)
    }
}